File: testes/algoritmos-ordenacao/js/quick-sort.js
export function quickSort(vetor) {
    let tamanhoVetor = vetor.length;

    if (tamanhoVetor <= 1) {
        return vetor;
    }

    const pivo = vetor[tamanhoVetor - 1];
    const left = [];
    const right = [];

    for (let i = 0; i < (tamanhoVetor-1); i++) {
        if (vetor[i] < pivo) {
            left.push(vetor[i]);
        } else {
            right.push(vetor[i]);
        }
    }

    return [...quickSort(left), pivo, ...quickSort(right)];
}

export function executarQuickSort(numeroRepeticoes, vetor, tipo_instancia) {
    let somaTempos = 0;
    let somaConsumosMemoria = 0;
    let mediaTempos = 0;
    let mediaConsumosMemoria = 0;
    let temposQuickSort = [];
    let consumosMemoriaQuickSort = [];

    for (let i = 0; i < numeroRepeticoes; i++) {
        let tempoInicial = 0;
        let tempoFinal = 0;
        let memoriaInicial = 0;
        let memoriaFinal = 0;

        memoriaInicial = process.memoryUsage().heapUsed;
        tempoInicial = performance.now().toFixed(6);
        quickSort(vetor);
        tempoFinal = performance.now().toFixed(6);
        memoriaFinal = process.memoryUsage().heapUsed;

        temposQuickSort.push((tempoFinal - tempoInicial));
        consumosMemoriaQuickSort.push((memoriaFinal - memoriaInicial));
    }

    //Cálculo da média de tempo gasto
    for (let i = 0; i < temposQuickSort.length; i++) {
        somaTempos += temposQuickSort[i];
    }
    mediaTempos = (somaTempos / temposQuickSort.length);

    //Cálculo da média de consumo de memória
    for (let i = 0; i < consumosMemoriaQuickSort.length; i++) {
        somaConsumosMemoria += consumosMemoriaQuickSort[i];
    }
    mediaConsumosMemoria = (somaConsumosMemoria / consumosMemoriaQuickSort.length);

    //Log de saída
    console.log(`
        Algoritmo: Quick Sort
        Tipo da Instância/Tamanho: ${tipo_instancia}/${vetor.length}
        Qtd. de Iterações: ${numeroRepeticoes}
        Média de Tempo Gasto: ${mediaTempos.toFixed(6)} ms
        Média de Consumo de Memória: ${mediaConsumosMemoria.toFixed(4)} bytes
    `);
}

File: testes/algoritmos-ordenacao/python/quick-sort.py
def quick_sort(vetor):
    tamanho_vetor = len(vetor)

    if tamanho_vetor <= 1:
        return vetor

    pivo = vetor[-1]
    left = []
    right = []

    for i in range(tamanho_vetor - 1):
        if vetor[i] < pivo:
            left.append(vetor[i])
        else:
            right.append(vetor[i])

    return quick_sort(left) + [pivo] + quick_sort(right)

def executar_quick_sort(numero_repeticoes, vetor, tipo_instancia):
    soma_tempos = 0
    soma_consumos_memoria = 0
    media_tempos = 0
    media_consumos_memoria = 0
    tempos_quick_sort = []
    consumos_memoria_quick_sort = []

    for i in range(numero_repeticoes):
        tempo_inicial = 0
        tempo_final = 0
        memoria_inicial = 0
        memoria_final = 0

        memoria_inicial = sys.getsizeof(vetor)
        tempo_inicial = time.time()
        quick_sort(vetor.copy())
        tempo_final = time.time()
        memoria_final = sys.getsizeof(vetor)

        tempos_quick_sort.append((tempo_final - tempo_inicial) * 1000)  # Em milissegundos
        consumos_memoria_quick_sort.append(memoria_final - memoria_inicial)

    soma_tempos = sum(tempos_quick_sort)
    media_tempos = soma_tempos / numero_repeticoes

    soma_consumos_memoria = sum(consumos_memoria_quick_sort)
    media_consumos_memoria = soma_consumos_memoria / numero_repeticoes

    print(f"""
        Algoritmo: Quick Sort
        Tipo da Instância/Tamanho: {tipo_instancia}/{len(vetor)}
        Qtd. de Iterações: {numero_repeticoes}
        Média de Tempo Gasto: {media_tempos:.6f} ms
        Média de Consumo de Memória: {media_consumos_memoria:.4f} bytes
    """)
